src/buffer/buffer_map.rs

```diff
@@ -24,27 +24,66 @@ use smallvec::SmallVec;
 use bevy_ecs::prelude::{Entity, World};
 
 use crate::{
-    add_listener_to_source, Accessed, AddOperation, AnyBuffer, AnyBufferKey, AnyMessageBox, Buffer,
-    BufferKeyBuilder, Bufferable, Buffered, Builder, Chain, Gate, GateState, Join, Joined,
-    OperationError, OperationResult, OperationRoster, Output, UnusedTarget,
+    add_listener_to_source, Accessed, AddOperation, AnyBuffer, AnyBufferKey, AnyMessageBox,
+    AsAnyBuffer, Buffer, BufferKeyBuilder, Bufferable, Buffered, Builder, Chain, Gate, GateState,
+    Join, Joined, OperationError, OperationResult, OperationRoster, Output, UnusedTarget,
 };
 
 pub use bevy_impulse_derive::JoinedValue;
 
-#[derive(Clone, Default)]
-pub struct BufferMap {
-    inner: HashMap<Cow<'static, str>, AnyBuffer>,
+/// Uniquely identify a buffer within a buffer map, either by name or by an
+/// index value.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum BufferIdentifier<'a> {
+    /// Identify a buffer by name
+    Name(Cow<'a, str>),
+    /// Identify a buffer by an index value
+    Index(usize),
 }
 
-impl BufferMap {
-    /// Insert a named buffer into the map.
-    pub fn insert(&mut self, name: impl Into<Cow<'static, str>>, buffer: impl Into<AnyBuffer>) {
-        self.inner.insert(name.into(), buffer.into());
+impl BufferIdentifier<'static> {
+    /// Clone a name to use as an identifier.
+    pub fn clone_name(name: &str) -> Self {
+        BufferIdentifier::Name(Cow::Owned(name.to_owned()))
     }
 
-    /// Get one of the buffers from the map by its name.
-    pub fn get(&self, name: &str) -> Option<&AnyBuffer> {
-        self.inner.get(name)
+    /// Borrow a string literal name to use as an identifier.
+    pub fn literal_name(name: &'static str) -> Self {
+        BufferIdentifier::Name(Cow::Borrowed(name))
+    }
+
+    /// Use an index as an identifier.
+    pub fn index(index: usize) -> Self {
+        BufferIdentifier::Index(index)
+    }
+}
+
+impl From<&'static str> for BufferIdentifier<'static> {
+    fn from(value: &'static str) -> Self {
+        BufferIdentifier::Name(Cow::Borrowed(value))
+    }
+}
+
+pub type BufferMap = HashMap<BufferIdentifier<'static>, AnyBuffer>;
+
+/// Extension trait that makes it more convenient to insert buffers into a [`BufferMap`].
+pub trait AddBufferToMap {
+    /// Convenience function for inserting items into a [`BufferMap`]. This
+    /// automatically takes care of converting the types.
+    fn insert_buffer<I: Into<BufferIdentifier<'static>>, B: AsAnyBuffer>(
```

Comment:
I think a potential problem with this is that index based keys are not guaranteed to be sequential, I don't know if this is good or not in practice, but I think we can change `insert_buffer` to only accept the string variant, and have a `extend_buffer` that takes an iterator and inserts the index keys sequentially.

I also thought of making `BufferMap` an enum

```rust
pub enum BufferMap<'a> {
  String(HashMap<Cow<'a, str>, AnyBuffer>),
  Index(Vec<AnyBuffer>),
}
```

But I guess there is probably a reason for the current impl.

Reply:
My expectation is that no one will do ordered iteration using this key. If someone wants to iterate over indices then they will `.get(&BufferIdentifier::Index(i))`.

Alternatively they may match the key to `BufferIdentifier::Index(index)` and then use `index` to align with the layout.

Reply:
> But I guess there is probably a reason for the current impl.

Mostly I found that funneling everything into the key made it much easier to have a simple and consistent error message. This also theoretically supports maps that can have a mix of names and indices but I don't think we're likely to need that.